feat(tree): staged allowed types (#24631)
microsoft/FluidFramework: .changeset/curly-bikes-train.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
"__section": feature
---
Adds enablable allowed types to SchemaFactoryAlpha

This adds the `enablable` API to [`SchemaFactoryAlpha`](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class). Enablables can be passed in as [`AllowedTypes`](https://fluidframework.com/docs/api/fluid-framework/allowedtypes-typealias) to declare that a type can be read at a particular location in the tree but not written there.

> This example array schema permits loading strings from the document but does not permit writing strings to the tree:

Reviewer: You should document this feature in terms of how it will help users, not here is an esoteric useless sounding thing you can do.

From this description many important things are unclear to me, even with the additional background I have on this feature.

Both the doc comment on SchemaFactoryAlpha.enablable and this changeset should make it clear that this feature allows schema evolution which adds a member to an AllowedTypes while supporting cross version collaboration.

Additionally they should explain how to use this feature to accomplish that.

Right now I can't tell how you enable something. I can't tell if this PR is intended to add some parts of the needed functionality, or the whole thing.

IIRC we consider ~3 possible designs for when you can write:

1. you can't write the type until the use of enablable is removed from the schema definition.
2. you can write the type if the underlying document stored schema supports it. Writes are allowed on unhydrated nodes but they will error on hydration if unsupported.
3. you can write the type, and the schema will be upgraded if you do (I don't think this is a good design, but its an option)

THen we also had a few options for how to handle the typeing:

1. make the types allow the writes, and use runtime errors only
2. make the types ban the writes (also use runtime errors if write is unallowed, maybe always, maybe just if not allowed by stored schema)
3. Provide some conditional typing schema (like some token you can get from doing the schema upgrade to allow them somehow)

And a few different ways for how the actual schema upgrade could work:

1. Each enableable type has a token associated with it which can be passed to the schemaUpgrade method on TreeView to enable it, updating the stored schema accordingly (and no op if already enabled). Also provide an API to check if a given enableable is already enabled. What UpgradeSchema with no arguments should do is unclear in this case: we would need to decide and document if it enables all enableables, or none. I'd suggest none, but maybe adding an option to enable all of them would make sese.
2. Update on first write of an enabable (again I don't think this is a good approach, but its possible)
3. No APi change and SchemaUpgrade enables all enablables (again I don't think this is a good approach)

From the documentation it should be clear which of these approaches is being taken, and how a user can follow the evolution end to end. It should also be clear if this is all supposed to be working after this change, or if key parts of this work are not yet finished.

Author: I'm resolving this because I think I've addressed it and it's pretty large so making it hard for me to keep track of what I'm doing. Feel free to unresolve if there's anything else :)

---
"fluid-framework": minor
"@fluidframework/tree": minor
"__section": feature
---
Adds enablable allowed types to SchemaFactoryAlpha

This adds the `enablable` API to [`SchemaFactoryAlpha`](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class).
Enablables can be used for schema evolution to add members to an [`AllowedTypes`](https://fluidframework.com/docs/api/fluid-framework/allowedtypes-typealias) while supporting cross version collaboration.

> Enablables are allowed types that can be read by the tree but not written. Attempting to write an enablable type will cause a runtime error.

Reviewer: This need more clarity. If I have one client which has finished the migration (did a schema upgrade) and one client which still has the enableable wrapper, what happens?

Also what does "write" mean? In this cross version collab case, can I modify an existing value if the enableable type? You can right?

Can I close a subtree that contains it and insert that clone?

Assuming that the ability to insert the type in that location is controlled by the current state of the stored schema, I think we should adopt phrasing more like:

```suggestion
Enablables are allowed types that can be enabled by schema upgrades.
Before being enabled, any attempt to insert or move a node to a location which requires the enablement for its type to be valid will throw an error.
```

Author: I like the updated phrasing a lot. I was struggling to make this sound correct :)

>Can I close a subtree that contains it and insert that clone?

I haven't thought about this scenario and I'm not sure if I understand it. What does closing a subtree and cloning it look like in the code?

Reviewer: If someone does `mynode.child = TreeBeta.clone(myNode.child);` you might expect it to replace the tree with an identical one, not error if it contains new stuff.

Same for if someone does `mynode.child = TreeAlpha.importVerbose(TreeAlpha.exportVerbose(myNode.child);`

It seems odd if such cases error. It sounds like the intention is for them to error, which is ok for now, but we should make it clear that this approach can cause such errors.

> To add a new member to an `AllowedTypes`, add the type wrapped by `enablable`:

Reviewer: Since the text after this example is discussing this hypothetical upgrade, we should introduce it as a hypothetical upgrade:

```suggestion
To add a new member to an `AllowedTypes`, add the type wrapped by `enablable`.
For example, migrating an array which previously supported only numbers to support both numbers and strings would start by deploying a version of the app using `enablable`:
```

```typescript
schemaFactoryAlpha.arrayAlpha("TestArray", [schemaFactoryAlpha.number, schemaFactoryAlpha.enablable(schemaFactoryAlpha.string)]);
```

Once enough clients have this code update, it is safe to allow writing strings to the array.
To enable writing strings to the array, a code change must be made to remove the enablable annotation:
```typescript

> schemaFactoryAlpha.arrayAlpha("TestArray", [schemaFactoryAlpha.number, schemaFactoryAlpha.string]);

Reviewer: I like the inclusion of an example. I think we should also have an integration test showing this feature end to end in actual code (something like what's in src/test/feature-libraries/modular-schema/schemaEvolutionExamples.spec.ts except targeting the public API, and less horrible).

Making an executable example as a test, then including the same example in the changeset or doc comments is a great pattern to help ensure everything works as promised.

I'd like to see a test with some end to end proof involving a schema upgrade showing the whole process working. If thats too big for the changeset, it can have a simplified version, but link to the full exampe code via a github link.

```

> A future change will add an API that allows enablables to be enabled via a runtime schema upgrade so that the type can be enabled using a configuration flag change rather than a code change.

Reviewer: Somone might read this thinking a future change refers to a future change of an app (like our example above). To clarify:

```suggestion
In the future, SharedTree may add an API that allows enablables to be enabled via a runtime schema upgrade so that the type can be more easily deployed using a configuration flag change rather than a code change.
```

---
"fluid-framework": minor
"@fluidframework/tree": minor
"__section": feature
---
Adds staged allowed types to SchemaFactoryAlpha

This adds the `staged` API to [`SchemaFactoryAlpha`](https://fluidframework.com/docs/api/fluid-framework/schemafactoryalpha-class).
Staged allowed types can be used for schema evolution to add members to an [`AllowedTypes`](https://fluidframework.com/docs/api/fluid-framework/allowedtypes-typealias) while supporting cross version collaboration.

> Staged allowed types are allowed types that can be upgraded by schema upgrades.

Reviewer: ```suggestion
Staged allowed types are [allowed types](https://fluidframework.com/docs/api/fluid-framework/allowedtypes-typealias) that can be upgraded by [schema upgrades}(https://fluidframework.com/docs/api/fluid-framework/treeview-interface#upgradeschema-methodsignature).
```

Before being upgraded, any attempt to insert or move a node to a location which requires its type to be upgraded to be valid will throw an error.

> To enable this feature, schema validation is now turned on by default except during initialization of the tree.

Reviewer: ```suggestion
To enable this feature, [schema validation](https://fluidframework.com/docs/api/fluid-framework/treeviewconfiguration-class#enableschemavalidation-property) is now performed by default when editing the tree.
```

To add a new member to an `AllowedTypes`, add the type wrapped by `staged`.
For example, migrating an array which previously supported only numbers to support both numbers and strings would start by deploying a version of the app using `staged`:
```typescript
schemaFactoryAlpha.arrayAlpha("TestArray", [schemaFactoryAlpha.number, schemaFactoryAlpha.staged(schemaFactoryAlpha.string)]);
```

Once enough clients have this code update, it is safe to allow writing strings to the array.
To allow writing strings to the array, a code change must be made to remove the staged annotation:
```typescript
schemaFactoryAlpha.arrayAlpha("TestArray", [schemaFactoryAlpha.number, schemaFactoryAlpha.string]);
```

> And then a schema upgrade is done to upgrade the stored schema:

Reviewer: Especially since the type of view is not clear from the code here, linking the docs would be good.

```suggestion
Then when opening old documents [upgradeSchema](https://fluidframework.com/docs/api/fluid-framework/treeview-interface#upgradeschema-methodsignature) is used to upgrade the stored schema:
```

```typescript
view.upgradeSchema()
```

> In the future, SharedTree will add an API that allows staged allowed types to be upgraded via a runtime schema upgrade so that the type can be more easily deployed using a configuration flag change rather than a code change.

Reviewer: Lets avoid making promises about future work. It's possible it will end up not being important enough to do, or not happen soon.

```suggestion
In the future, SharedTree may add an API that allows staged allowed types to be upgraded via a runtime schema upgrade so that the type can be more easily deployed using a configuration flag change rather than a code change.
```

> Below is a full example of how the schema migration process works. This can also be found in our tests.

Reviewer: Any app which claims to really be a "full" example here needs to subscribe to schema change events and handle the associated invalidation from a schema upgrade which might make it no longer viewable. Maybe worth linking TreeViewComponent, https://github.com/microsoft/FluidFramework/blob/aaf6dd747f105c300d656a31e7f0db43c4b2450d/experimental/framework/tree-react-api/src/reactSharedTreeView.tsx#L250 as a reference for how to do that if this examples doesn't.

```typescript
// schema A: only number allowed
const schemaA = factory.optional([SchemaFactoryAlpha.number]);

// schema B: number or string (string is staged)
const schemaB = factory.optional([
	SchemaFactoryAlpha.number,
	factory.staged(SchemaFactoryAlpha.string),
]);

// schema C: number or string, both fully allowed
const schemaC = factory.optional([SchemaFactoryAlpha.number, SchemaFactoryAlpha.string]);

> const provider = new TestTreeProviderLite(3);

Reviewer: Advertising to customers that how you write simple  examples for how to use our public API surface is to use a non exported private test utility from the tree package is not great.

I think we should try and do this using the actual public API surface. If that means this version of the code has to be in the end-to-end tests instead of the tree package, or that it has to use independentView, then it can be updated accordingly.
